File: AddTwoNumbers/AddTwoNumbers.dart
// Definition for singly-linked list.
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    ListNode dummy = new ListNode(0);
    ListNode? l3 = dummy;
    int carry = 0;

    while (l1 != null || l2 != null) {
      int val = ((l1?.val ?? 0) + (l2?.val ?? 0) + carry);
      l3?.next = new ListNode(val % 10);
      if (l1 != null) {
        l1 = l1.next;
      }
      if (l2 != null) {
        l2 = l2.next;
      }
      carry = (val / 10).toInt();
      l3 = l3?.next;
    }

    if (carry == 1) {
      l3?.next = new ListNode(1);
    }

    return dummy.next;
  }
}

File: AddTwoNumbers/AddTwoNumbers.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l3 = ListNode(0)
        head = l3
        carry = 0
        while l1 is not None or l2 is not None:
            first = 0
            second = 0
            if l1 is not None:
                first = l1.val
                l1 = l1.next
            if l2 is not None:
                second = l2.val
                l2 = l2.next
            temp = first + second + carry
            carry = 1 if temp >= 10 else 0
            l3.next = ListNode(temp % 10)
            l3 = l3.next
        if carry == 1:
            l3.next = ListNode(1)
        return head.next
            

    def addTwoNumbersOld(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = 0
        digit = 0
        while l1 is not None or l2 is not None:
            first = 0
            second = 0
            if l1 is not None:
                first = l1.val
                l1 = l1.next
            if l2 is not None:
                second = l2.val
                l2 = l2.next
            result += (first + second) * 10**digit
            digit += 1
        return result